### DISCUSSION TITLE
Support for exposing Cache-Control-related headers in response

### DISCUSSION THREAD

**Discussion by andrew-demb** [nginxinc/nginx-s3-gateway#227] [Q&A] [answered]
If the file is found and available, I want to generate responses with the added `Cache-Control` header (for example: `Cache-Control: public, max-age=31536000, immutable`).
The added cache headers allow to cache of responses by CDN (like Cloudflare https://developers.cloudflare.com/cache/concepts/cache-control/).

I can try to achieve this with additional reverse proxy Nginx and the `add_header` function (https://nginx.org/en/docs/http/ngx_http_headers_module.html). 
But this solution will expose cache-related headers about "safe to cache", even if the response was unsuccessful.

Does `nginx-s3-gateway` support exposing cache-related headers?

----

**UPD**: after another round of reading Nginx docs, I found that `expires` function can help me for such use-case https://nginx.org/en/docs/http/ngx_http_headers_module.html#expires
But I have a bit more complex use-case, than described above.
I generate response from my PHP application with added HTTP headers (`Cache-Control`) and using internal forwarding to `nginx-s3-gateway` with `X-Accel-Redirect` - https://www.nginx.com/resources/wiki/start/topics/examples/x-accel/
In such use-case I cannot rely over Nginx `expires` function :(

**Comment by 4141done** [comment, marked as answer]
Hello @andrew-demb  - very sorry for the delay in my response.  Thank you for providing so much detail in your question.

My understanding is that you want to generate your own unique `Cache-Control` headers in the response from the s3 gateway.  You do not want to pass on headers from S3, correct?

In any case, we include a blank configuration file(https://github.com/nginxinc/nginx-s3-gateway/blob/master/common/etc/nginx/templates/gateway/s3_server.conf.template) that you can override to inject your own configuration. Instructions for how to do that are here https://github.com/nginxinc/nginx-s3-gateway/blob/master/docs/development.md#stub-files

In that file, if you add something like 
```nginx
add_header "Cache-Control" "public, max-age=31536000, immutable";
```

You can I think control things however you'd like. I have not personally tested this out but it might be worth a try to see if it works for your use case. 

Happy to work through it with you if this does not work or I misunderstood your situation.  I'll admit that I didn't fully understand issue you noted in your update to this question.

**Reply by andrew-demb** [reply]
Thank you for the answer.
I think it would work in general and will be helpful to mention this way of proxy configuration in readme.

In our case, we had a bit of misconfigured Cloudflare in addition to the provided details  - Cloudflare cached response if it doesn't contain the header "Cache-Control" too (refs: [1][2]).
So if we cannot change Cloudflare settings - if we will use something like `add_header "Cache-Control" "public, max-age=31536000, immutable";` - response won't contain `Cache-Control` header.

As of our Nginx reverse proxy with internal location (to be used only by X-Accel-Redirect) already have embedded Lua support - we decided to use Lua + `error_page` to override header `Cache-Control` in case of error status code.

```
# ...
    location ~ ^/multimedia/(.+) {
        internal;

        # ...
        proxy_intercept_errors on;
        error_page 400 404 500 503 = @uncached_multimedia;

        proxy_pass http://media-proxy/$1;
    }

    # Allows you to ensure that caching is disabled for multimedia unsuccessful responses
    location @uncached_multimedia {
        header_filter_by_lua_block {
            ngx.header["Cache-Control"] = "no-cache";
        }
    }
# ...
```

[1] https://developers.cloudflare.com/cache/concepts/default-cache-behavior/
[2] https://developers.cloudflare.com/cache/how-to/edge-browser-cache-ttl/set-browser-ttl/
